File: tests/f90/test_kernel_negate.f90
! Track B conformance fixture: unary minus (flang `Negate`). Covers the bare
! leaf form (-y), a compound operand needing printer parentheses (-2.0*x(i) —
! Fortran parses this as the negation of the whole term, R1008), and negation
! of a source-parenthesized expression (-(x + y)).

module kernel_negate_mod
  implicit none
contains

  subroutine neg_clip(x, y, n)
    integer,            intent(in)    :: n
    real, dimension(n), intent(in)    :: x
    real, dimension(n), intent(inout) :: y
    integer :: i

    do concurrent (i=1:n)
      if (x(i) < -y(i)) then
        y(i) = -2.0*x(i)
      else
        y(i) = -(x(i) + y(i))
      endif
    enddo

  end subroutine neg_clip

end module kernel_negate_mod

File: tests/cpp/test_kernel_negate.cpp
// Track B conformance fixture (C++ side): unary minus — the mirror of
// tests/f90/test_kernel_negate, with one deliberate asymmetry it exists to
// pin: C++ unary minus binds TIGHTER than `*`, so `-2.0_rt * x` parses as
// `(-2.0_rt) * x` (Neg on the literal), unlike Fortran's R1008 where
// `-2.0*x(i)` negates the whole term. Also covers the bare leaf (-y) and
// negation of a source-parenthesized sum (-(x + y)).

using Real = double;
constexpr Real operator""_rt(long double v) { return static_cast<Real>(v); }

void neg_clip_point(Real& y, Real const x) noexcept
{
    if (x < -y) {
        y = -2.0_rt * x;
    } else {
        y = -(x + y);
    }
}
